src/Lean/Meta/Tactic/AC/Main.lean

```diff
@@ -170,21 +159,55 @@ where
       | none => return Simp.Step.done { expr := e }
     | e, _ => return Simp.Step.done { expr := e }
 
-def rewriteUnnormalizedRefl (goal : MVarId) : MetaM Unit := do
-  let newGoal ← rewriteUnnormalized goal
-  newGoal.refl
+def rewriteUnnormalized (mvarId : MVarId) : MetaM MVarId := do
+  let simpCtx :=
+    {
+      simpTheorems  := {}
+      congrTheorems := (← getSimpCongrTheorems)
+      config        := Simp.neutralConfig
+    }
+  let tgt ← instantiateMVars (← mvarId.getType)
+  let (res, _) ← Simp.main tgt simpCtx (methods := { post })
+  applySimpResultToTarget mvarId tgt res
 
-def rewriteUnnormalizedNormalForm (goal : MVarId) : TacticM Unit := do
-  let newGoal ← rewriteUnnormalized goal
-  replaceMainGoal [newGoal]
+def rewriteUnnormalizedRefl (goal : MVarId) : MetaM Unit := do
+  (← rewriteUnnormalized goal).refl
 
 @[builtin_tactic acRfl] def acRflTactic : Lean.Elab.Tactic.Tactic := fun _ => do
   let goal ← getMainGoal
   goal.withContext <| rewriteUnnormalizedRefl goal
 
-@[builtin_tactic acNf] def acNfTactic : Lean.Elab.Tactic.Tactic := fun _ => do
-  let goal ← getMainGoal
-  goal.withContext <| rewriteUnnormalizedNormalForm goal
+/-- Implementation of the `ac_nf` tactic when operating on the main goal. -/
+def acNfTarget : TacticM Unit :=
+  liftMetaTactic1 fun goal => do return (← rewriteUnnormalized goal)
+
+/-- Implementation of the `ac_nf` tactic when operating on a hypothesis. -/
+def acNfHyp (fvarId : FVarId) : TacticM Unit :=
```

Comment:
Can we generalize this to `acNfExpr (e : Expr) : TacticM Unit` instead? Then we can write `acNFHyp` as a thin wrapper around this. Happy to do this if you want me to @tobiasgrosser .